### ISSUE TITLE
fix_shape() in use_tensorrt_safe_mode is called differently

### ISSUE THREAD

**Issue by ravijo** [haotian-liu/yolact_edge#59]
Hi,

Thank you for developing this amazing tool.

I noticed that the function `fix_shape` is used differently in the following two places:

`master` branch [detection.py#L194](https://github.com/haotian-liu/yolact_edge/blob/3aa5e39db1ed1a057baf79c2e1a74debc7ee83d8/layers/functions/detection.py#L194)
```
        try:
            classes = classes[keep]
            boxes = boxes[keep]
            masks = masks[keep]
            scores = scores[keep]
        except IndexError:
            from utils.logging_helper import log_once
            log_once(self, "issue_27_flatten", name="yolact.layers.detect", message="Encountered IndexError as mentioned in https://github.com/haotian-liu/yolact_edge/issues/27. Flattening predictions to avoid error, please verify the outputs. If there are any problems you met related to this, please report an issue.")

            keep = torch.flatten(keep, end_dim=1)
            idx = torch.nonzero(keep, as_tuple=True)[0]

            classes, boxes, masks, scores = [flatten_index_select(x, idx, end_dim=1)
                                             for x in (classes, boxes, masks, scores)]

        classes, boxes, masks, scores = fix_shape(classes, boxes, masks, scores)
```
Please notice that `fix_shape` is outside the `except` block. Therefore, if there is no `IndexError`, the `fix_shape` will be still be called. Now please see the second use of `fix_shape()`

`haotian-dev` branch [detection.py#L220](
https://github.com/haotian-liu/yolact_edge/blob/c28ddb6d2d130046828106d275069905e2460d54/layers/functions/detection.py#L220)
```
            keep = torch.flatten(keep, end_dim=1)
            idx = torch.nonzero(keep, as_tuple=True)[0]

            classes, boxes, masks, scores = [flatten_index_select(x, idx, end_dim=1)
                                             for x in (classes, boxes, masks, scores)]
            classes, boxes, masks, scores = fix_shape(classes, boxes, masks, scores)
```
Please notice that in this case, `fix_shape` is being called after `flatten_index_select`.

I am wondering if I am missing something. Please have a look.

**Comment by haotian-liu** [comment]
`fix_shape` is implemented to solve TensorRT engine related issue, thus we only need it when we want to resolve all TRT issue -- the TensorRT Safe Mode.

**Comment by ravijo** [comment]
Yes. You are right. 

Please allow me to rephrase myself. I am trying to say that [here in detection.py#L191-L194](https://github.com/haotian-liu/yolact_edge/blob/3aa5e39db1ed1a057baf79c2e1a74debc7ee83d8/layers/functions/detection.py#L191-L194), `fix_shape` should be a part of the `except` block. Right now, in the `master` branch, the `fix_shape`  is called even if there is no `IndexError`.

**Comment by haotian-liu** [comment]
You might notice that earlier, there is no `fix_shape` function. I implemented this to fix TensorRT engine related issue, and I also noticed that sometimes the IndexError won't be triggered but there was still the issue. That's why I moved this out.

But as I have previously said, it is only needed to resolve TRT issue, thus in my dev branch, I placed it under the if statement instead of applying it to all cases.

**Comment by ravijo** [comment]
I agree.

https://github.com/haotian-liu/yolact_edge/blob/3aa5e39db1ed1a057baf79c2e1a74debc7ee83d8/layers/functions/detection.py#L179-L194

The above code is taken from the `master` branch. Please pay attention to line number 194. Now, as per your recent addition of `use_tensorrt_safe_mode` in `haotian-dev` branch, we can see the following extension :

https://github.com/haotian-liu/yolact_edge/blob/c28ddb6d2d130046828106d275069905e2460d54/layers/functions/detection.py#L209-L220

The above code is taken from the `haotian-dev` branch. Please pay attention on how `fix_shape` is called in two different ways (i.e., `master` branch vs `haotian-dev` branch). 

In conclusion, I think that in the `master` branch, at line number 194, there should be 4 extra spaces so that `fix_shape` is called only in case of `IndexError` after `flatten_index_select`. _(I have not verified my claim, so please ignore if needed)_

**Comment by haotian-liu** [comment]
I might not have been clear enough. `fix_shape` is introduced to resolve a case that we see when TensorRT engine is enable, but it does not necessarily trigger a IndexError. Thus, it should be apply to all the case so that we handle all those cases. However, we do not see that happening when TensorRT is not enabled. Thus we only apply to one branch for TensorRT-Safe-Mode. Hope this clarifies your doubts.
